p2p-specs/p2p-interface.md

```diff
@@ -0,0 +1,229 @@
+
+# Bundler interaction domains
+
+## Configuration
+
+This section outlines constants that are used in this spec.
+
+| Name | Value | Description |
+|---|---|---|
+
+
+## MetaData
+
+Bundlers MUST locally store the following `MetaData`:
+
+```
+(
+  
+)
+```
+
+Where
+
+- .
+- .
+
+
+## The gossip domain: gossipsub
+
+Bundlers MUST support the [gossipsub v1](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.0.md) libp2p Protocol
+including the [gossipsub v1.1](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.1.md) extension.
+
+### Topics and messages
+
+Topics are plain UTF-8 strings and are encoded on the wire as determined by protobuf (gossipsub messages are enveloped in protobuf messages).
+Topic strings have form: `/erc4337/UserOpsWithEntryPoint/Name/Encoding`.
+This defines both the type of data being sent on the topic and how the data field of the message is encoded.
+
+- `UserOpsWithEntryPoint` - the lowercase hex-encoded (no "0x" prefix) bytes of `entry_point_contract_address` + `user_operation` 
+
+Each gossipsub [message](https://github.com/libp2p/go-libp2p-pubsub/blob/master/pb/rpc.proto#L17-L24) has a maximum size of `GOSSIP_MAX_SIZE`.
+Bundlers MUST reject (fail validation) messages that are over this size limit.
+Likewise, Bundlers MUST NOT emit or propagate messages larger than this limit.
+
+The `message-id` of a gossipsub message MUST be the following 20 byte value computed from the message data:
+* If `message.data` has a valid snappy decompression, set `message-id` to the first 20 bytes of the `SHA256` hash of
+  the concatenation of `MESSAGE_DOMAIN_VALID_SNAPPY` with the snappy decompressed message data,
+  i.e. `SHA256(MESSAGE_DOMAIN_VALID_SNAPPY + snappy_decompress(message.data))[:20]`.
+* Otherwise, set `message-id` to the first 20 bytes of the `SHA256` hash of
+  the concatenation of `MESSAGE_DOMAIN_INVALID_SNAPPY` with the raw message data,
+  i.e. `SHA256(MESSAGE_DOMAIN_INVALID_SNAPPY + message.data)[:20]`.
+
+*Note*: The above logic handles two exceptional cases:
+(1) multiple snappy `data` can decompress to the same value,
+and (2) some message `data` can fail to snappy decompress altogether.
+
+The payload is carried in the `data` field of a gossipsub message, and varies depending on the topic:
+
+| Name                             | Message Type              |
+|----------------------------------|---------------------------|
+| `UserOperationWithEntryPoint`    | ``                        |
+| `NewPooledUserOperationsHashes`  | ``                        |
+
+
+Bundlers MUST reject (fail validation) messages containing an incorrect type, or invalid payload.
+
+When processing incoming gossip, Bundlers MAY de-score or disconnect peers who fail to observe these constraints.
+
+For any optional queueing, Bundlers SHOULD maintain maximum queue sizes to avoid DoS vectors.
+
+
+#### Global topics
```

Comment:
I think we're still missing a way for bundlers to advertise which mempool-ids they support (or maybe a list of mempool-ids should be a part of the request).

Reply:
Can we include the list of mempool-ids supported by a node to be part of the metadata? @yoavw

Reply:
Yes